pypsa/statistics.py

```diff
@@ -190,6 +204,48 @@ def get_transmission_carriers(n, bus_carrier=None):
     )
 
 
+# TODO: likely remove
+def fallback_at_port(groupby, at_port):
+    """
+    Check whether ports should be taken into account when grouping assets.
+
+    This is decided by the groupby argument that is used. If it is a function
+    that contains a keyword argument `port`, the ports will be taken into account,
+    else not.
+    """
+    ports_available = callable(groupby) and "port" in signature(groupby).parameters
+    if at_port and not ports_available:
+        logger.warning(
+            "Argument 'groupby' is not a function that supports "
+            "'port' as a argument. Falling back to not using ports."
+        )
+        at_port = False
+    return at_port
+
+
+def get_grouping(n, c, groupby, port=None, nice_names=False) -> [pd.Series, list]:
+    by = None
+    level = None
+    if callable(groupby):
+        if "port" in signature(groupby).parameters:
+            by = groupby(n, c, port=port, nice_names=nice_names)
+        else:
+            by = groupby(n, c, nice_names=nice_names)
+    elif isinstance(groupby, list):
+        by = [n.df(c)[key] for key in groupby]
+    elif isinstance(groupby, str):
+        by = n.df(c)[groupby]
+    elif isinstance(groupby, dict):
+        # TODO: fix kwargs behavior. what is this used for again?
```

Comment:
I think this was formerly used for the energy balance, where we wanted group by the index levels so we needed something like {"level":["bus", "carrier", "bus_carrier"]}

Reply:
I think we can now remove it